plugins/cpu/fu-cpu-device.c

```diff
@@ -382,6 +425,250 @@ fu_cpu_device_add_security_attrs (FuDevice *device, FuSecurityAttrs *attrs)
+do_fu_cpu_wr_fw_cleanup (GError **error, const gchar *msg, FuDevice *device)
+{
+	if (msg) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     msg);
+	}
+	fu_device_set_status (device, FWUPD_STATUS_IDLE);
+}
+
+#define fu_cpu_wr_fw_cleanup(msg)	do_fu_cpu_wr_fw_cleanup(error, msg, device)
+
+static gboolean
+fu_cpu_device_write_firmware (FuDevice *device, FuFirmware *firmware,
+				FwupdInstallFlags flags, GError **error)
+{
+	const gchar *ucode_dir = "/lib/firmware/intel-ucode";
+	FuCpuDevice *self = FU_CPU_DEVICE (device);
+	g_autoptr(GOutputStream) ostream = NULL;
+	g_autofree gchar *ucode_bkp_path = NULL;
+	g_autofree gchar *ucode_path = NULL;
+	g_autoptr(GBytes) ucode_blob = NULL;
+	GFile* ucode_fd;
+	gssize size;
+
+	/* how do we set fw version using MSR plugin ? */
+
+	/* check device updatable */
+	if (!fu_cpu_is_updatable (error)) {
+		fu_cpu_wr_fw_cleanup ("device not updatable");
+		return FALSE;
+	}
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_VERIFY);
+
+	if (firmware == NULL) {
+		fu_cpu_wr_fw_cleanup ("firmware is NULL");
+		return FALSE;
+	}
+	ucode_blob = fu_firmware_get_bytes (firmware, error);
+	if (ucode_blob == NULL) {
+		fu_cpu_wr_fw_cleanup ("firmware blob is NULL");
+		return FALSE;
+	}
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+
+	if (g_getenv ("FWUPD_CPU_VERBOSE") != NULL)
+		fu_common_dump_bytes (G_LOG_DOMAIN, "microcode", ucode_blob);
+
+	ucode_path = g_strdup_printf ("%s/%02x-%02x-%02x", ucode_dir,
+					self->family_id, self->model_id,
+					self->stepping_id);
+	ucode_bkp_path = g_strdup_printf ("%s%s", ucode_path, UCODE_BKP_EXTN);
+
+	if (ucode_path == NULL || ucode_bkp_path == NULL) {
+		fu_cpu_wr_fw_cleanup ("failed to allocate memory");
+		return FALSE;
+	}
+
+	/* only do update if we can make backup of existing microcode */
+	if (g_file_test (ucode_path, G_FILE_TEST_EXISTS)) {
+		if (g_rename (ucode_path, ucode_bkp_path) < 0) {
+			fu_cpu_wr_fw_cleanup ("failed to create backup of existing microcode");
+			return FALSE;
+		}
+	}
+
+	ucode_fd = g_file_new_for_path (ucode_path);
+	ostream = G_OUTPUT_STREAM (g_file_replace (ucode_fd, NULL, FALSE,
+					G_FILE_CREATE_NONE, NULL, error));
+
+	if (ostream == NULL) {
+		fu_cpu_wr_fw_cleanup ("failed to create ucode file");
+		/* restore previous microcode */
+		if (!fu_cpu_restore_microcode (ucode_path))
+			g_prefix_error (error, "failed to restore previous microcode\n");
+		return FALSE;
+	}
+
+	size = g_output_stream_write_bytes (ostream, ucode_blob, NULL, error);
+	if (!g_output_stream_close (ostream, NULL, error)) {
+		fu_cpu_wr_fw_cleanup ("failed to close ostream");
+		if (!fu_cpu_restore_microcode (ucode_path))
+			g_prefix_error (error, "failed to restore previous microcode\n");
+		return FALSE;
+	}
+	if (size < 0) {
+		fu_cpu_wr_fw_cleanup ("failed to write ucode file");
+		if (!fu_cpu_restore_microcode (ucode_path))
+			g_prefix_error (error, "failed to restore previous microcode\n");
```

Comment:
for all of your `g_prefix_error` calls, you should have them end in `: ` instead of `\n`